### PR TITLE
feat(data type): ArrayType support inner dataType

### LINKED ISSUE #4974: Feature: Merge VariantArrayType into ArrayType
**Summary**

Description for this feature.

1. Merge VariantArrayType into ArrayType
2. Support `Get`, `MapAccess` for `ArrayColumn`

So we should indicate the inner type, eg:

```
col_name Array(Varchar)
``` 

If the inner type is variant:

```
col_name Array(JSON)
```

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

ArrayType support inner dataType

## Changelog

- New Feature
- Build/Testing/CI
- Documentation

## Related Issues

Fixes #4974

### CODE REVIEW

File: common/datavalues/src/array_value.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+use std::fmt;
+use std::fmt::Debug;
+use std::fmt::Display;
+use std::fmt::Formatter;
+
+use crate::prelude::*;
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Default, Clone, PartialEq)]
+pub struct ArrayValue {
+    pub values: Vec<DataValue>,
```
Comment (sundy-li) [inline common/datavalues/src/array_value.rs:27]:
```
pub struct ArrayValue {
     column: ColumnRef,
     offset:  usize,
     length: usize,
}
```

I would like to use this struct, it's just a slice of the inner column.

File: tests/suites/0_stateless/03_dml/03_0023_insert_into_array.sql [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+select '==Array(String)==';
+
+CREATE TABLE IF NOT EXISTS t12(id Int, arr Array(String)) Engine = Memory;
+
+INSERT INTO t12 (id, arr) VALUES(1, ['aa', 'bb']), (2, ['cc', 'dd']);
+
+select * from t12;
+
+select '==Array(String) Nullable==';
```
Comment (sundy-li) [inline tests/suites/0_stateless/03_dml/03_0023_insert_into_array.sql:101]: I think we don't accept `Array(String null)`, so tests should be covered about this type.

Comment (b41sh) [reply tests/suites/0_stateless/03_dml/03_0023_insert_into_array.sql:101]: ok

**Comment by sundy-li** [comment]
@mergify update

**Comment by BohuTANG** [comment]
Oops, CI failure:
```
--- /workspace/tests/suites/0_stateless/03_dml/03_0023_insert_into_array.result	2022-05-07 07:40:38.903043700 +0000
+++ /workspace/tests/suites/0_stateless/03_dml/03_0023_insert_into_array.stdout	2022-05-07 07:46:59.932615348 +0000
@@ -1,3 +1,4 @@
+ERROR 1105 (HY000) at line 129: Code: 1005, displayText = sql parser error: Expected ), found: Null.
 ==Array(UInt8)==
 1	[1, 2, 3]
 2	[254, 255]
```
https://github.com/datafuselabs/databend/runs/6333086848?check_suite_focus=true#step:3:437

File: tests/suites/0_stateless/03_dml/03_0023_insert_into_array.sql (hunk 2 above)
Comment (b41sh) [reply tests/suites/0_stateless/03_dml/03_0023_insert_into_array.sql]:
```SQL
CREATE TABLE IF NOT EXISTS t16(id Int, arr Array(String Null)) Engine = Memory; -- {ErrorCode 1005}
```
will parse error, we can't test this case in the stateless test.

**Review by sundy-li — APPROVED** [review]

**Comment by sundy-li** [comment]
This may be a conflict with #5167, cc @zhang2014

**Pull request merged**